File: day1/day1.cpp
#include <stdio.h>
#include <vector>
#include <algorithm>

bool read_vectors(std::vector<int> &l1, std::vector<int> &l2)
{
    FILE* f = fopen("input2.txt", "r");
    if (f == NULL)
    {
        return false;
    }

    int i = 0;
    int j = 0;
    while (fscanf(f, "%d,%d\n", &i, &j) != EOF)
    {
        l1.push_back(i);
        l2.push_back(j);
    }
    return true;
}

long unsigned int find_list_difference(std::vector<int> &l1, 
        std::vector<int> &l2)
{
    long unsigned int total_dist = 0;
    long unsigned int c = 0;
    while (c < l1.size())
    {
        total_dist += abs(l1[c] - l2[c]);
        c++; // HEY!
    }

    return total_dist;
}

long unsigned int find_list_similarity(std::vector<int> &l1, 
        std::vector<int> &l2)
{
    long unsigned int total_similarity = 0;

    for (long unsigned int i = 0; i < l1.size(); i++)
    {
        int appearances = 0;
        for (long unsigned int j = 0; j < l1.size(); j++)
        {
            if (l1[i] == l2[j])
                appearances++;
        }
        total_similarity += appearances * l1[i];
    }
    return total_similarity;
}

int main()
{
    std::vector<int> list1;
    std::vector<int> list2;
    read_vectors(list1, list2);

    if (list1.size() != 0)
    {
        printf("Good size of list\n");
    }

    
    std::sort(list1.begin(), list1.end());
    std::sort(list2.begin(), list2.end());
    long unsigned int d = find_list_difference(list1, list2);
    long unsigned int s = find_list_similarity(list1, list2);
    
    printf("List difference: %ld\n", d);
    printf("List similarity: %ld\n", s);
    return 0;
}

File: day1/day1.py
"""
Takes two lists of equal length, l1 and l2, and returns the distance between.
"""
def find_list_difference(l1: list, l2: list) -> int:
    assert len(l1) == len(l2)

    total_dist = 0
    c = 0
    while c < len(l1):
        total_dist += abs(l1[c] - l2[c])
        c += 1

    return total_dist


def find_list_similarity(l1: list, l2: list) -> int:
    assert len(l1) == len(l2)

    total_similarity = 0
    for i in l1:
        appearances = 0
        for j in l2:
            if (i == j):
                appearances += 1
        total_similarity += appearances * i

    return total_similarity

f = open("input.txt", "r")

list1 = []
list2 = []

for line in f:
    nums = line.split("   ")
    list1.append(int(nums[0]))
    list2.append(int(nums[1]))

list1.sort()
list2.sort()


test1 = [3, 4, 2, 1, 3, 3]
test2 = [4, 3, 5, 3, 9, 3]

test1.sort()
test2.sort()

print("Total distance found, test: " + str(find_list_difference(test1, test2)))
print("Total distance found, true: " + str(find_list_difference(list1, list2)))

print("Total similarity found,test: " + str(find_list_similarity(test1, test2)))
print("Total similarity found,true: " + str(find_list_similarity(list1, list2)))
